core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -93,16 +96,77 @@ private static StaticDataTask.Row convertPartition(Partition partition) {
     return StaticDataTask.Row.of(partition.key, partition.recordCount, partition.fileCount, partition.specId);
   }
 
-  private static Iterable<Partition> partitions(StaticTableScan scan) {
+  private static Iterable<Partition> partitions(Table table, StaticTableScan scan) {
     CloseableIterable<FileScanTask> tasks = planFiles(scan);
+    Types.StructType normalizedPartitionType = Partitioning.partitionType(table);
+    PartitionMap partitions = new PartitionMap();
+
+    LoadingCache<Integer, Integer[]> originalPartitionFieldPositionsBySpec = Caffeine.newBuilder().build(specId ->
+        originalPositions(table, specId, normalizedPartitionType));
+    LoadingCache<Pair<PartitionData, Integer>, PartitionData> normalizedPartitions = Caffeine.newBuilder().build(
+        pData -> normalizePartition(pData, normalizedPartitionType, originalPartitionFieldPositionsBySpec));
 
-    PartitionMap partitions = new PartitionMap(scan.table().spec().partitionType());
     for (FileScanTask task : tasks) {
-      partitions.get(task.file().partition()).update(task.file());
+      PartitionData original = (PartitionData) task.file().partition();
+      int specId = task.spec().specId();
+      PartitionData normalized = normalizedPartitions.get(Pair.of(original, specId));
+      partitions.get(normalized).update(task.file());
     }
     return partitions.all();
   }
 
+  /**
+   * Returns an array of original partition field positions, indexed by normalized partition field positions
+   */
+  private static Integer[] originalPositions(Table table,
```

Comment:
I think if we make this non-static we can just rely on the table coming from this class

Reply:
I see this is all called from a static context so maybe we can't change that ...